docker/docker_client.go

```diff
@@ -419,6 +404,30 @@ func (c *dockerClient) makeRequest(ctx context.Context, method, path string, hea
 // makeRequest should generally be preferred.
 // TODO(runcom): too many arguments here, use a struct
 func (c *dockerClient) makeRequestToResolvedURL(ctx context.Context, method, url string, headers map[string][]string, stream io.Reader, streamLen int64, auth sendAuth, extraScope *authScope) (*http.Response, error) {
+	var (
+		res   *http.Response
+		err   error
+		delay time.Duration
+	)
+	// Do an exponential back off starting at 2 seconds for at most 5
+	// iterations.
+	delay = 2
+	for i := 0; i < 5; i++ {
+		res, err = c.makeRequestToResolvedURLHelper(ctx, method, url, headers, stream, streamLen, auth, extraScope)
+		if res != nil && res.StatusCode == http.StatusTooManyRequests {
+			logrus.Debugf("%s: sleeping for %d seconds before next attempt", http.StatusText(res.StatusCode), delay)
+			if i < 4 {
+				time.Sleep(delay * time.Second)
+				delay *= 2
+			}
+			continue
+		}
+		break
+	}
+	return res, err
+}
+
+func (c *dockerClient) makeRequestToResolvedURLHelper(ctx context.Context, method, url string, headers map[string][]string, stream io.Reader, streamLen int64, auth sendAuth, extraScope *authScope) (*http.Response, error) {
```

Comment:
`Helper` says nothing, especially without any documentation about its purpose. `…WithoutRetries`, maybe?

Reply:
Added a comment and renamed it to `makeRequestToResolvedURLOnce`.